### PR TITLE
ui: Add organization and user account settings (PROJQUAY-4553)

### PR DESCRIPTION
**Pull request by jonathankingfc**
- This PR adds settings pages to the organization and user organization pages.
- Admin users can edit their preferences, billing, and organization type
- Updated cypress version to address bug https://github.com/cypress-io/cypress/issues/25397

### CODE REVIEW

File: web/src/routes/OrganizationsList/Organization/Tabs/Settings/BillingInformation.tsx [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+          />
+          <Radio
+            id="radio-organization"
+            label="Organization Account"
+            name="radio-organization"
+            onClick={() => setAccountType('organization')}
+            isChecked={accountType == 'organization'}
+            description="Multiple users and teams that share access and billing under a single namespace"
+          />
+          {user.organizations.length > 0 && accountType == 'organization' && (
```
Comment (bcaton85) [inline web/src/routes/OrganizationsList/Organization/Tabs/Settings/BillingInformation.tsx:301]: The view details still appears for a user that doesn't belong to any organizations.

File: web/src/routes/OrganizationsList/Organization/Tabs/Settings/GeneralSettings.tsx [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+          tag_expiration_s:
+            getSeconds(timeMachineFormValue) != namespaceTimeMachineExpiry
+              ? getSeconds(timeMachineFormValue)
+              : null,
+          email: namespaceEmail != emailFormValue ? emailFormValue : null,
+          isUser: isUserOrganization,
+        });
+        return response;
+      } else {
+        const response = await updateUser({
```
Comment (bcaton85) [inline web/src/routes/OrganizationsList/Organization/Tabs/Settings/GeneralSettings.tsx:199]: Is this missing time machine settings for the user? Currently setting the time machine for user namespaces doesn't do anything. Org namespaces still work.

File: web/src/routes/OrganizationsList/Organization/Tabs/Settings/GeneralSettings.tsx [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+import {useOrganization} from 'src/hooks/UseOrganization';
+import {useOrganizationSettings} from 'src/hooks/UseOrganizationSettings';
+import {IOrganization} from 'src/resources/OrganizationResource';
+import {humanizeTimeForExpiry, getSeconds, isValidEmail} from 'src/libs/utils';
+import {addDisplayError} from 'src/resources/ErrorHandling';
+import {useQuayConfig} from 'src/hooks/UseQuayConfig';
+import {useUpdateUser} from 'src/hooks/UseCurrentUser';
+
+type validate = 'success' | 'warning' | 'error' | 'default';
+const timeMachineOptions = {
```
Comment (bcaton85) [inline web/src/routes/OrganizationsList/Organization/Tabs/Settings/GeneralSettings.tsx:27]: Custom tag options set by `TAG_EXPIRATION_OPTIONS` doesn't seem to be supported. For example if I set `TAG_EXPIRATION_OPTIONS` to `2d, 5d, 6w` nothing get's rendered in the list

File: web/src/routes/OrganizationsList/Organization/Tabs/Settings/BillingInformation.tsx [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+      <Flex
+        spaceItems={{default: 'spaceItemsSm'}}
+        direction={{default: 'column'}}
+      >
+        <FlexItem>
+          <Title headingLevel="h3">
+            {currentPlan?.plan.toUpperCase()} organization&apos;s plan
+          </Title>
+        </FlexItem>
+        {privateAllowed ? (
```
Comment (bcaton85) [inline web/src/routes/OrganizationsList/Organization/Tabs/Settings/BillingInformation.tsx:216]: nit: there's the `Conditional` component that may make this easier to read. More of a matter of preference. 
```
<Conditional if={privateAllowed}>
    <FlexItem>{`${privateCount} of ${privateAllowed} private repositories used`}</FlexItem>
</Conditional>
```

File: web/src/hooks/UseOrganizationSettings.ts [hunk 5]
```diff
@@ -1,23 +1,16 @@
...
-      email,
-      isUser,
-    }: updateOrgSettingsParams): Promise<Response> => {
-      return await updateOrgSettings(
-        namespace,
-        tag_expiration_s,
-        email,
-        isUser,
-      );
+    async (params: Partial<updateOrgSettingsParams>): Promise<Response> => {
```
Comment (bcaton85) [inline web/src/hooks/UseOrganizationSettings.ts:12]: Wasn't aware of `Partial`, great use.

**Review by bcaton85 — CHANGES_REQUESTED** [review]
Great work, just some small bugs that need to be fixed.  I wasn't able to test the billing features, will follow up asking for instructions directly. 

Also there's other settings I saw in the old UI that I didn't find in the new UI. I'm assuming these will be done in future PR's? Do we have JIRA's for these?
- Change user password
- Application tokens
- Enable/disable desktop notifications
- Delete account/organization

**Comment by dmage** [comment]
@jonathankingfc fyi, I cherry-pick your changes for cypress into https://github.com/quay/quay/pull/2171 so that we can backport cypress without backporting the feature if needed

**Comment by Sunandadadi** [comment]
Discussed changes through slack.

File: web/src/routes/RepositoriesList/RobotAccountsList.tsx [hunk 6]
```diff
@@ -159,6 +159,7 @@ export default function RobotAccountsList(props: RobotAccountsListProps) {
           props.organizationName,
         ]);
       },
+      enabled: !props.isUser,
```
Comment (Sunandadadi) [inline web/src/routes/RepositoriesList/RobotAccountsList.tsx:162]: This is already there on line 155

File: web/src/routes/OrganizationsList/Organization/Tabs/Settings/Settings.tsx [hunk 7]
```diff
@@ -239,13 +21,21 @@ export default function Settings(props: SettingsProps) {
...
-    // {
-    //   name: 'Billing Information',
-    //   id: 'billinginformation',
-    //   content: <BillingInformation />,
-    // },
+    {
+      name: 'Billing Information',
+      id: 'billinginformation',
+      content: <BillingInformation organizationName={props.organizationName} />,
+    },
```
[the remark this answers is not in the document]
Comment (Sunandadadi) [reply web/src/routes/OrganizationsList/Organization/Tabs/Settings/Settings.tsx:288]: Can add `visible` key here? Something like [this](https://github.com/quay/quay/pull/2151/files#diff-e9a6c92a9f0242880ababdad54680c0dddca82d57a05ef44a564e8629108d388R282)

File: web/src/routes/OrganizationsList/Organization/Organization.tsx [hunk 8]
```diff
@@ -114,7 +119,7 @@ export default function Organization() {
           isUser={isUserOrganization}
         />
       ),
-      visible: true,
+      visible: fetchTabVisibility('Robot accounts'),
```
Comment (Sunandadadi) [inline web/src/routes/OrganizationsList/Organization/Organization.tsx:122]: Is there a reason why you are changing this?

**Review by Sunandadadi — CHANGES_REQUESTED** [review]

File: local-dev/stack/config.yaml [hunk 9]
```diff
@@ -76,6 +76,7 @@ CORS_ORIGIN:
   - "https://stage.foo.redhat.com:1337"
   - "http://localhost:9000"
 FEATURE_UI_V2: True
+FEATURE_USER_METADATA: True
```
Comment (dmage) [inline local-dev/stack/config.yaml:79]: What is this used for?

File: web/cypress/fixtures/config.json [hunk 10]
```diff
@@ -1,146 +1,146 @@
 {
-    "account_recovery_mode": false,
```
Comment (dmage) [inline web/cypress/fixtures/config.json:2]: If you want to change how this file is formatted, please do it in a separate PR. It's unclear what is changed in this file.

File: web/src/components/modals/GenerateEncryptedPasswordModal.tsx [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+            onChange={(value) => setPassword(value)}
+            aria-label="text input example"
+            label="Password"
+          />
+          Please enter your password in order to generate
+        </>
+      )}
+      {step == 2 && (
+        <>
+          Your encrypted password is: <br /> {clientKey}
```
[the remark this answers is not in the document]
Comment (Sunandadadi) [reply web/src/components/modals/GenerateEncryptedPasswordModal.tsx:70]: +1 
This is helpful. Can we add this?

File: web/src/routes/OrganizationsList/Organization/Organization.tsx (hunk 8 above)
Comment (jonathankingfc) [reply web/src/routes/OrganizationsList/Organization/Organization.tsx:122]: This should only be available for specific permissions

File: web/cypress/fixtures/config.json (hunk 10 above)
Comment (jonathankingfc) [reply web/cypress/fixtures/config.json]: The pre-commit checks are failing if I do not format this file

File: web/src/routes/OrganizationsList/Organization/Tabs/Settings/Settings.tsx [hunk 12]
```diff
@@ -266,6 +26,15 @@ export default function Settings(props: SettingsProps) {
...
     },
     {
+      name: 'Billing Information',
+      id: 'billinginformation',
+      content: <BillingInformation organizationName={props.organizationName} />,
+      visible: quayConfig?.features?.BILLING,
+    },
+  ];
+
+  if (quayConfig?.features?.AUTO_PRUNE) {
```
Comment (Sunandadadi) [inline web/src/routes/OrganizationsList/Organization/Tabs/Settings/Settings.tsx:36]: The `if` condition here is not required as visible prop [here](https://github.com/quay/quay/pull/2151/files#diff-e9a6c92a9f0242880ababdad54680c0dddca82d57a05ef44a564e8629108d388R46) does the same.

File: web/src/routes/OrganizationsList/Organization/Tabs/Settings/GeneralSettings.tsx [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+  const organizationName = props.organizationName;
+  const {user, loading: isUserLoading} = useCurrentUser();
+  const {organization, isUserOrganization, loading} =
+    useOrganization(organizationName);
+  const [error, setError] = useState<string>('');
+
+  const {updateOrgSettings} = useOrganizationSettings({
+    name: organizationName,
+    onSuccess: (result) => {
+      setAlerts((prevAlerts) => {
```
Comment (Sunandadadi) [inline web/src/routes/OrganizationsList/Organization/Tabs/Settings/GeneralSettings.tsx:53]: You can use `useAlerts` hook. Eg: https://github.com/quay/quay/blob/master/web/src/routes/OrganizationsList/Organization/Tabs/TeamsAndMembership/TeamsView/TeamViewKebab.tsx#L25C7-L25C15

Comment (Sunandadadi) [reply web/src/routes/OrganizationsList/Organization/Tabs/Settings/GeneralSettings.tsx:53]: Similar for other places where you are setting Alerts.

**Review by Sunandadadi — CHANGES_REQUESTED** [review]

**Comment by Sunandadadi** [comment]
Nice job, a few other comments:

1. I was not able to save users General Settings when the fields (Full Name / company / Location) are entered. An API call is not triggered here.
2. Can you combine the various hooks files into `Billing`, `CLI`, `General Settings` hook? @harishsurf pointed out that having multiple files for hooks makes it difficult to find the operations. Trying to have one file for one view. You can make this change in a follow up PR if that works.

File: web/src/routes/OrganizationsList/Organization/Tabs/Settings/Settings.tsx (hunk 12 above)
Comment (jonathankingfc) [reply web/src/routes/OrganizationsList/Organization/Tabs/Settings/Settings.tsx]: Resolved

**Comment by jonathankingfc** [comment]
@Sunandadadi Thanks for the review, for those fields to work you need to enable FEATURE_USER_METADATA: true in the config.yaml. We mock this config value in the e2e tets

**Comment by Sunandadadi** [comment]
> @Sunandadadi Thanks for the review, for those fields to work you need to enable FEATURE_USER_METADATA: true in the config.yaml. We mock this config value in the e2e tets

Got it, in that case can we hide them and display when `FEATURE_USER_METADATA` is true? It is confusing to have them but not be able to update.

File: web/src/routes/OrganizationsList/Organization/Tabs/Settings/GeneralSettings.tsx (hunk 13 above)
Comment (jonathankingfc) [reply web/src/routes/OrganizationsList/Organization/Tabs/Settings/GeneralSettings.tsx]: Updated to use useAlerts hook

**Comment by jonathankingfc** [comment]
> > @Sunandadadi Thanks for the review, for those fields to work you need to enable FEATURE_USER_METADATA: true in the config.yaml. We mock this config value in the e2e tets
> 
> Got it, in that case can we hide them and display when `FEATURE_USER_METADATA` is true? It is confusing to have them but not be able to update.

Updated to hide when FEATURE_USER_METADATA is false

**Review by Sunandadadi — APPROVED** [review]
lgtm
